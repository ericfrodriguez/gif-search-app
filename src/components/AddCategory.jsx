import { useState } from 'react'


export const AddCategory = ({ onNewCategory }) => {

    const [inputValue, setInputValue] = useState('')

    const handleInputChange = ({ target }) => {
        setInputValue(target.value);
    }

    const handleSubmit = (event) => {
        event.preventDefault();

        if (inputValue.trim().length <= 1) return;

        onNewCategory(inputValue.trim());

        setInputValue('');
    }

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                placeholder="Buscar GIFs"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}