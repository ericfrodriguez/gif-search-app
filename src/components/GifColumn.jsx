import { GifCard } from "./index";

export const GifColumn = ({ images }) => {

    return (
        <div className='column'>
            {
                images.map(image => (
                    <GifCard 
                        key={image.id}
                        {...image}
                    />
                ))
            }
        </div>
    )
}