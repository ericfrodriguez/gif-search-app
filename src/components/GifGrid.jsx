import { GifColumn } from "./index";
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from "prop-types";

export const GifGrid = ({ category }) => {

    const { images, isLoading } = useFetchGifs(category);

    return (
        <>
            <h3>{category}</h3>
            <div className="row">
                {
                    isLoading && <h2>Cargando...</h2>
                }
                {
                    images.map((column, index) => (
                        <GifColumn
                            key={index}
                            images={column}
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

export default GifGrid