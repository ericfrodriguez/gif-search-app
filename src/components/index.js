
export * from './AddCategory'
export * from './GifGrid'
export * from './GifCard'
export * from './GifColumn'