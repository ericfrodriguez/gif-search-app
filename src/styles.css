@import url('https://fonts.googleapis.com/css2?family=Poppins:wght@500&display=swap');

* {
    font-family: Helvetica, Arial, sans-serif;
    background-color: rgb(234, 234, 234);
    box-sizing: border-box;
}

body {
    padding: 60px;
}

input {
    background-color: white;
    border-radius: 5px;
    border: 1px solid rgb(97, 32, 158);
    color: black;
    font-size: 1.2rem;
    outline: none;
    padding: 10px 15px;
    width: 100%;
}

h2 {
    font-size: 1.5rem;
}

h3 {
    font-size: 1.8rem;
    margin-bottom: 5px;
}

.row {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
}

.column {
    flex: 100%;
    max-width: 100%;
    padding: 0 4px;
}

.card {
    position: relative;
    align-content: center;
    align-items: center;
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-top: 8px;
    background-color: white;
    box-shadow: 0px 3px 5px rgba(0, 0, 0, 0.03);
    overflow: hidden;
    vertical-align: middle;
    transition: box-shadow .2s ease-in-out;
}

.card:hover {
    -webkit-box-shadow: 4px 4px 0px 0px rgba(124, 15, 207, 1);
    -moz-box-shadow: 4px 4px 0px 0px rgba(124, 15, 207, 1);
    box-shadow: 4px 4px 0px 0px rgba(124, 15, 207, 1);
}

.card img {
    width: 100%;
    overflow: hidden;
    transition: transform .2s ease-in-out;
}

.card:hover img {
    transform: scale(1.4);
}

.card img:hover~p {
    display: block;
}


.card p {
    opacity: 0;
    display: flex;
    justify-content: center;
    align-items: flex-end;
    position: absolute;
    padding: 0 20px 10px 20px;
    width: 100%;
    height: 100%;
    margin: 0;
    font-family: 'Poppins', sans-serif;
    font-size: 1rem;
    text-align: center;
    color: white;
    background: #280046;
    background: linear-gradient(0deg, rgba(40, 0, 70, 1) 0%, rgba(45, 157, 253, 0) 73%);
    transition: opacity .2s ease-in-out;
}

.card p:hover {
    opacity: 100;
    backdrop-filter: brightness(40%);
}

@media screen and (min-width: 600px) {
    .column {
        flex: 50%;
        max-width: 50%;
    }

    .card p {
        font-size: 1.2rem;
    }

    h3 {
        font-size: 2.4rem;
    }
}

@media screen and (min-width: 1000px) {
    .column {
        flex: 25%;
        max-width: 25%;
    }

    .card p {
        font-size: 1.5rem;
    }

    h3 {
        font-size: 3rem;
    }
}