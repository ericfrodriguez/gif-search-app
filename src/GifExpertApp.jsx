import { useState } from "react"
import { AddCategory, GifGrid} from "./components";


const GifExpertApp = () => {
    const [categories, setCategories] = useState(['One Punch']);

    const categoryExists = (newCategory) => {
        const matchCategory = new RegExp(newCategory, 'i');

        return categories.some(cat => matchCategory.test(cat))
    }

    /**
     * @param {String} newCategory New category to render
     */
    const handleAddCategory = (newCategory) => {


        if (categoryExists(newCategory)) return;

        setCategories([newCategory, ...categories])
    }

    return (
        <>
            <h1>Gif Expert</h1>
            <AddCategory
                onNewCategory={handleAddCategory}
            />
            {
                categories.map((category) => (
                    <GifGrid
                        key={category}
                        category={category}
                    />
                ))
            }
        </>
    )
}

export default GifExpertApp