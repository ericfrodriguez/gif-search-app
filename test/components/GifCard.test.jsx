import { render, screen } from "@testing-library/react";
import { GifCard } from "../../src/components";

describe('Tests on <GifCard />', () => {

    const GifCardsProps = {
        title: 'Saitama GIF',
        url: 'https://media1.giphy.com/media/yo3TC0yeHd53G/200w.gif?cid=261d21f9kx4b2tr4hhxofjbyco998kgrq6b7jt28m6igzklf&rid=200w.gif&ct=g'
    }

    test('Should match the snapshot', () => {


        const { container } = render(<GifCard {...GifCardsProps} />);

        expect(container).toMatchSnapshot();
    })

    test('It should show the image and the indicated ALT', () => {

        render(<GifCard {...GifCardsProps} />)

        const { alt, src } = screen.getByRole("img");

        expect(src).toBe(GifCardsProps.url)
        expect(alt).toBe(GifCardsProps.title)

    })

    test('It should display the title in the component', () => {

        render(<GifCard {...GifCardsProps} />)

        expect(screen.getByText(GifCardsProps.title)).toBeTruthy();

    })
})