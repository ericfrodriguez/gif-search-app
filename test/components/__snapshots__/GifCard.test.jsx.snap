// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <GifCard /> Should match the snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Saitama GIF"
      src="https://media1.giphy.com/media/yo3TC0yeHd53G/200w.gif?cid=261d21f9kx4b2tr4hhxofjbyco998kgrq6b7jt28m6igzklf&rid=200w.gif&ct=g"
    />
    <p>
      Saitama GIF
    </p>
  </div>
</div>
`;
